python/tank/descriptor/bundle_cache_usage/manager.py

```diff
@@ -0,0 +1,270 @@
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import os
+import time
+import threading
+
+from worker import BundleCacheUsageWorker
+from exception import BundleCacheUsageException
+from exception import BundleCacheUsageFileDeletionException
+from . import BundleCacheUsageLogger as log
+
+
+class BundleCacheManager(object):
+
+    """
+    Bungle utility class for scanning, purging a bundle packages in the bundle cache
+    """
+
+    # keeps track of the single instance of the class
+    __singleton_lock = threading.Lock()
+    __singleton_instance = None
+
+    def __new__(cls, *args, **kwargs):
+        #
+        # Based on tornado.ioloop.IOLoop.instance() approach.
+        #
+        # See:
+        #   https://github.com/facebook/tornado
+        #   https://gist.github.com/werediver/4396488
+        #   https://en.wikipedia.org/wiki/Double-checked_locking
+        #
+        if not cls.__singleton_instance:
+            log.debug_manager("__new__")
+            with cls.__singleton_lock:
+                if not cls.__singleton_instance:
+                    cls.__singleton_instance = super(BundleCacheManager, cls).__new__(cls, *args, **kwargs)
+                    cls.__singleton_instance.__initialized = False
+
+        return cls.__singleton_instance
+
+    def __init__(self, bundle_cache_root):
+        super(BundleCacheManager, self).__init__()
+        log.debug_manager("__init__('%s')" % bundle_cache_root)
+        # TODO: returning would cause a silent non-usage of specified parameter
+        if (self.__initialized):
+            return
+        self._worker = None
+
+        if bundle_cache_root is None:
+            raise ValueError("The 'bundle_cache_root' parameter is None.")
+
+        if not os.path.exists(bundle_cache_root):
+            raise ValueError("The specified 'bundle_cache_root' parameter folder does not exists.")
+
+        if not os.path.isdir(bundle_cache_root):
+            raise ValueError("The specified 'bundle_cache_root' parameter is not a directory.")
+
+        self._bundle_cache_root = bundle_cache_root
+        self._worker = BundleCacheUsageWorker(bundle_cache_root)
+        self._worker.start()
+        self.__initialized = True
+
+    @classmethod
+    def delete_instance(cls):
+        with cls.__singleton_lock:
+            if cls.__singleton_instance:
+                if cls.__singleton_instance._worker:
+                    cls.__singleton_instance._worker.stop()
+                    cls.__singleton_instance = None
+
+    @classmethod
+    def _find_app_store_path(cls, base_folder):
+        for (dirpath, dirnames, filenames) in os.walk(base_folder):
+            if dirpath.endswith('app_store'):
+                return dirpath
+
+        return None
+
+    @classmethod
+    def _find_descriptors(cls, base_folder, max_walk_depth):
+        path_list = []
+        base_folder_len = len(base_folder)
+        for (dirpath, dirnames, filenames) in os.walk(base_folder):
```

Comment:
Once again, why are we visiting the folder recursively? We have a pretty flat structure.

Reply:
Two nested `listdir` would get rid of the string manipulation, the max_walk_depth variable and the need to scan individual files inside the bundle.

```
path_list = []
for bundle_name in os.listdir(base_folder):
    bundle_folder = os.path.join(base_folder, bundle_name)
    for version in os.listdir(bundle_folder):
        version_path = os.path.join(bundle_folder, version):
        if os.path.exists(os.path.join(version_path,  "info.yml")):
            path_list.append(version_path)
```

is much simpler to understand and doesn't require string manipulation or counting slashes.

Reply:
This method by itself is super easy to understand what is going on.